src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectTracker.kt

```diff
@@ -10,8 +10,35 @@
 
 package org.jetbrains.kotlinx.lincheck.strategy.managed
 
+import sun.nio.ch.lincheck.WeakIdentityReference
+import org.jetbrains.kotlinx.lincheck.util.*
+import org.jetbrains.kotlinx.lincheck.transformation.isJavaLambdaClass
+import kotlin.coroutines.Continuation
+import java.lang.ref.ReferenceQueue
+import java.lang.ref.WeakReference
+import java.util.WeakHashMap
+
 /**
- * Tracks object allocations and changes in object graph topology.
+ * Tracks objects and changes in object graph topology.
+ *
+ * The object tracker allows registering objects within the managed strategy,
+ * assigning to each object a unique serial number and unique id.
+ *
+ * The unique serial number is a 32-bit integer number.
+ * Assuming the deterministic execution, the serial numbers are guaranteed to be generated in the same order,
+ * providing a persistent objects numeration between different re-runs of the same program execution.
+ *
+ * The unique object id is a 64-bit integer number, constructed from
+ * the object's serial number and its identity hash code.
+ *
+ * The registered objects are associated with the registry entries [ObjectEntry],
+ * keeping object's serial number, its identity hash code, a weak reference to the object,
+ * and, potentially, other meta-data (defined by the concrete implementations of the interface).
+ *
+ * The tracker allows retrieving the registry entry either by the object reference itself or by unique object id.
+ * In this way, the tracker allows to:
+ *   - get an object by its object id: `ObjectId -> Object`;
+ *   - get an object id by the object itself: `Object -> ObjectId`.
  */
 interface ObjectTracker {
```

Comment:
Is this interface necessary? Idem for the abstract class. I see only one actual implementation. Or are there plans to write more implementations for this interface and even more for the abstract class?

Reply:
Yes, I agree, it is over-engineering here. Let's keep it as is for now to merge it quicker, and revisit this later.